src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs

```diff
@@ -273,9 +259,13 @@ public override int Read(byte[] array, int offset, int count)
                 Debug.Assert(_inflater.NeedsInput(), "We can only run into this case if we are short of input");
 
                 int bytes = _stream.Read(_buffer, 0, _buffer.Length);
-                if (bytes == 0)
+                if (bytes <= 0)
                 {
-                    break;      //Do we want to throw an exception here?
+                    break;
+                }
+                else if (bytes > _buffer.Length)
+                {
+                    throw new InvalidDataException(SR.GenericInvalidData);
```

Comment:
Yes.  I'll add a comment.  With the implementation today, that'll end up throwing later on but almost by accident, and if we end up doing any perf work that would switch to using unsafe code, those checks could end up going away, hence I wanted to move it up here close to the site of where the bad data enters the system.